### ISSUE TITLE
Feature req: Spoof original source IP on internal connection

### ISSUE THREAD

**Issue by Steltek** [dlundquist/sniproxy#179]
Rather than try to pass the source IP via a header (which isn't possible without decrypting and mangling the HTTPS traffic), would it be possible to add an option to have sniproxy spoof the original source IP on the internal requests? That way, the target server could see the source IP.

(Of course, the user installing sniproxy would have to ensure it is 'in-path', i.e. all return traffic from the internal server gets funneled back to sniproxy for processing, and the OS allows for such spoofing.)

**Example:**
Client IP: 8.8.8.8
Gateway external IP: 80.80.80.80 (internalserver.domain.com)
Gateway internal IP: 10.0.0.1
Internal webserver IP: 10.0.0.2
- sniproxy installed on gateway.
1. TCP packet from client IP to 'internalserver.domain.com': src IP 8.8.8.8, dst IP 80.80.80.80 port 443
2. sniproxy opens an internal connection (on internal interface): src IP 8.8.8.8, dst IP 10.0.0.2 port 8443
3. Internal webserver responds: src IP 10.0.0.2, port 8443, dst IP 8.8.8.8
4. Packet arrives on GW which uses rule to redirect traffic from internal port 8443 to sniproxy
5. sniproxy sends response to Client via external interface: src IP 80.80.80.80 port 443, dst IP 8.8.8.8

-> The server thinks it is talking to 8.8.8.8 but in reality has all its return traffic intercepted by the gateway and sent to sniproxy instead.

References: haproxy's transparent mode:
http://blog.haproxy.com/2011/08/03/layer-7-load-balancing-transparent-proxy-mode
http://blog.haproxy.com/2012/06/05/preserve-source-ip-address-despite-reverse-proxies/

**Comment by akarl10** [comment]
Since I needed the same I implemented a proof of concept. If no source address is given in the configuration sniproxy tries IP_TRANSPARENT (in this quick and dirty try I always set this option on the listening socket, but this shoud no do any harm)

note that setting the flag requires root privileges, so if you like to test you must run it with "user root"

if needed I can also do a pull request with a cleaner method, but for now the patch as txt attachment.
[tproxy.txt](https://github.com/dlundquist/sniproxy/files/736902/tproxy.txt)

the usual iptables TPROXY setups should work.

**Comment by dlundquist** [comment]
@akarl10 You've convinced me this is a significantly invasive change. I've pushed a new branch containing this change enabled in the configuration by listener "source client". Would you mind testing this?

**Comment by akarl10** [comment]
looks good.

There is one thing that maybe should be documented as well:
looking around there is currently no way to set the resolv_mode on a single resolv_query call. That would be needed to enforce the same address family would be used on the outgoing socket (a minor problem easily resolved by using a instance for sniproxy for every address family).

according to man 7 ip IP_TRANSPARENT should also work with cap_net_admin, but I had no success setting this capability using setcap on the binary. Maybe this has to be set with cap_set_proc() call before changing to user. I will look into that next week

**Comment by akarl10** [comment]
I made some experiments running IP_TRANSPARENT without root.
Doing so requires a new depenency (libcap)
I discovered that the man page is wrong about the needed capability: CAP_NET_RAW is the required capability, not CAP_NET_ADMIN

[capabilities.txt](https://github.com/dlundquist/sniproxy/files/769196/capabilities.txt)

**Comment by dlundquist** [comment]
@akarl10 Thank you for the follow on patch. I would like to include a test script and autoconf feature switches so we can easily setup a test environment for this and so this doesn't break newer versions for non Linux users before merging this into master. Since there is a source address conflict when using IP_TRANSPARENT on local connections I was thinking of a test script (maybe modification of functional_test) which utilizes network namespaces and veth interface pairs to mock out a three host test setup. Thanks for including the autoconf config for libcap, that is a significant portion of the way there for not breaking non-Linux users. I haven't had much time to dedicate to sniproxy lately, but would like to see this one merged.

**Comment by fiforms** [comment]
Just want to say I'm very please with this new feature. I have a build from the "tproxy" branch running in production as we speak. I'm using "source client," and it does what I expected, allowing separate servers on a private LAN to share a single public IP, while not losing the ability to read REMOTE_ADDR from scripts on the internal servers. This is very useful. Thanks!

**Comment by dlundquist** [comment]
@fiforms Thank you for the positive report.

**Comment by dlundquist** [comment]
@akarl10 I've put together a functional test for transparent proxy operation using network namespaces, and found the `setsockopt(..., IP_TRANSPARENT, ...)` call was not needed on the listening socket. There is still some work in aligning capabilities and the binder process.

**Comment by akarl10** [comment]
@dlundquist I've made some comments on a commit but I cannot find the comments anymore (I never commented commits on github, maybe I made something wrong)

Yes, the listening socket does not need IP_TRANSPARENT,
the other thing I commented was that the only reason bind fails after setting IP_TRANSPARENT that succeeds is when running on the same host. That means that in connection.c you should ignore the case if bind fails when IP_TRANSPARENT succeeds. This would have the advantage that connections from the same host are possible with the loose of the "transparent" functionality, but the src address will be right since its local :)

**Comment by mrdomino** [comment]
Is it possible to use this with sniproxy and the downstream servers on the same machine without using network namespaces? Or if not, is it reasonable to use network namespaces in production?

**Comment by dlundquist** [comment]
@mrdomino The use of network namespaces was simply to reproduce a deployment pattern when the transparent proxy operates on the NAT gateway for testing, and not required for production deployments. That being said, I'm not sure transparent proxy can be accomplished to a local server since this would mean there are two open sockets to the same same remote address. The complication arises in how the replies from the backend servers are routed back through sniproxy, this depends on IPtables functionality that I'm not completely familiar with.

If not using the transparent proxy feature there is no problem running the sniproxy and the backend services on the same host.

Network namespaces are pretty mature in recent Linux kernels, although I would recommend a kernel newer than 3.13.

**Comment by mrdomino** [comment]
Right, that makes sense. I'm already running sniproxy + downstream services on a local box, so I'd have to make a shift to split them up. It seems like network namespaces are the way to do that.

**Comment by akarl10** [comment]
@mrdomino Since I needed the transparent proxy feature to proxy to systems "behind", but also to services running on the local box I did the following

I solved this using iptables (since I have practically no experience with namespaces):
1. change httpd (or nginx, whatever you are using) to another port (I used 444)
2. configure sniproxy to use port 443
3. setup iptables for example like this:

iptables -t mangle -N DIVERT
iptables -t mangle -A DIVERT -j MARK --set-mark 0x08/0x08
iptables -t mangle -A DIVERT -j ACCEPT

iptables -t mangle -A OUTPUT -p tcp --sport 444 -j MARK --set-mark 0x8/0x8

iptables -t mangle -I PREROUTING -p tcp -m socket -j DIVERT

ip -4 ru a pref 100 fwmark 0x8 lookup 100
ip -4 r a local default dev lo table 100 metric 50000 # metric higher than 1024 so that strongswan does not use 127.0.0.1 as outgoing address

This will not work out of the box for connections coming from the same host. For this to work the change I suggested in my previous comment would be needed.
